### ISSUE TITLE
DualStack: Add IPv4/IPv6 Dual-Stack support and awareness

### ISSUE THREAD

**Issue by leblancd** [kubernetes/kubernetes#62822]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
> /kind bug
/kind feature

**What happened**:
For Kubernetes Release 1.10 or older, clusters can only be run in either IPv4-only, IPv6-only, or in a VERY limited dual-stack configuration. The current dual-stack support is limited by the following restrictions:
- Pods can have dual-stack addresses (if a CNI plugin is used that supports dual-stack, e.g. bridge or calico), but Kubernetes is only aware of 1 address per pod
- Nodes can have dual-stack addresses
- Kube system pods (api server, controller manager, etc.) can only have 1 address per pod (all pods have IPv4 or all pods have IPv6)
- Only 1 service CIDR can be configured for the cluster, so service IPs are all IPv4 or all IPv6
- Kube-proxy only supports 1 family of iptables at a time (choose iptables or ip6tables)
- Endpoints for services are all IPv4 or all IPv6
- Kube-dns is capable of running dual-stack, but it is currently only made aware of either all IPv4 or all IPv6 service addresses

**What you expected to happen**:
Kubernetes should be aware of and support dual-stack addresses for pods, nodes and services. To make this happen, the required changes will include (but not be limited to) the following:
- Kubernetes needs to be made aware of multiple IPs per pod (up to one IPv4, possibly multiple IPv6 addrs)
- Kubernetes needs to be configurable for multiple (at least 2) service CIDRs
- Kube-proxy (and IPVS) needs to support IPv4 and IPv6 services in parallel (e.g. drive iptables and ip6tables in parallel)
- Health/liveness/readiness probes should be configurable for IPv4 or IPv6, with possibly IPv4 probes in parallel with IPv6 probes
- Cloud provider plugins should be able to report dual-stack external/internal addresses
- Kubectl commands and output displays will need to be modified for dual-stack
- Some Kubeadm support will probably need to be added for dual-stack
- New e2e test cases will probably need to be added to test parallel IPv4/IPv6 connectivity between pods, nodes, and services

Some questions to be answered when a feature spec is written:
- Should individual service manifests be able to select IPv4 or IPv6 service IPs, or every service gets IPv4 + IPv6 service IPs?
- How to extend dual-stack to future multiple interface per pod support? (Or at least how not to "paint  ourselves into a corner" with the dual-stack changes?)

**How to reproduce it (as minimally and precisely as possible)**:
Just try to set up a Kubernetes cluster in dual-stack mode.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): 1.10 or older
- Cloud provider or hardware configuration: Any
- OS (e.g. from /etc/os-release): Any
- Kernel (e.g. `uname -a`): Any
- Install tools:
- Others:

**Corresponding Feature Issue**:
https://github.com/kubernetes/features/issues/563

**Task/To-Do List**:
https://docs.google.com/spreadsheets/d/1KjvjvmZkhLI9M1BXVRVNn_oQpcwCB424x-0lOY2a3G0/edit#gid=0

**Comment by pmichali** [comment]
Two other factors to consider are maintaining backward compatibility with IPv4/IPv6 only configurations, and the impact on configuration files/templates.

Some questions...

What would be the benefit of doing **both** IPv4 and IPv6 liveness/readiness probes? Would either one or the other mode be adequate?

How will display be handled, in light of multiple IPv6 addresses (e.g. get pods)?

Should IPv4 be given precedence, when both are available?

Is the intent to "derive" the mode, based on IP addresses/CIDRs being used, rather than explicitly specifying the mode to use? I would think the former is more flexible.

**Comment by anguslees** [comment]
Agreed: given that a (wildcard) IPv6 listening socket can also serve IPv4 connections, I suggest we only need the one probe.  In other words: the probe is checking *application health*, and the application probably only has a single socket, whereas the specific IPv4/IPv6 pod addresses are CNI's problem and out of scope.  (Also out of scope are routing, firewalling, and all those other reasons why a container might be "healthy" and yet still unreachable)

Is the intent to generalise support for multiple pod addresses/CIDRs, or limit it only to what is required for multiple address families?  Specifically: do we intend to also support multiple pod/service CIDRs as part of this, or is that out of scope and we intend to special-case it to only a single IPv4 CIDR and a single IPv6 CIDR?

**Comment by leblancd** [comment]
Task/To-Do list:
https://docs.google.com/spreadsheets/d/1KjvjvmZkhLI9M1BXVRVNn_oQpcwCB424x-0lOY2a3G0/edit#gid=0

**Comment by pkanna000** [comment]
Is this enhancement request closed or implemented in 1.13 version?(Add IPv4/IPv6 Dual-Stack support )

**Comment by pmichali** [comment]
The KEP for this feature was moved from https://github.com/kubernetes/community/pull/2254 to https://github.com/kubernetes/enhancements/pull/648.

I'll let @leblancd comment on the status of the KEP, but for myself, I won't be able to continue working on dual-stack (I do have an outstanding PR out for review - https://github.com/kubernetes/kubernetes/pull/70659).

**Comment by aojea** [comment]
Update with the new KEP reference https://github.com/kubernetes/enhancements/pull/808

**Comment by athenabot** [comment]
@leblancd
If this issue has been triaged, please comment `/remove-triage unresolved`.

If you aren't able to handle this issue, consider unassigning yourself and/or adding the `help-wanted` label.

🤖 I am a bot run by vllry. 👩‍🔬

**Comment by pacoxu** [comment]
/remove-lifecycle rotten
It will be beta in 1.21 release.

**Comment by aojea** [comment]
> /remove-lifecycle rotten
> It will be beta in 1.21 release.
> [kubernetes/enhancements#563 (comment)](https://github.com/kubernetes/enhancements/issues/563#issuecomment-765737903)

I think that we can close the issue since is tracked as enhacement
/close
